src/persist-txn/src/txn_cache.rs

```diff
@@ -940,6 +968,71 @@ impl<T: Timestamp + TotalOrder> DataTimes<T> {
     }
 }
 
+#[derive(Debug)]
+pub(crate) enum Unapplied<'a> {
+    Batch(&'a Vec<u8>),
+    Register,
+}
+
+struct UnappliedIter<'a, T>
+where
+    T: Timestamp + Lattice + TotalOrder + StepForward + Codec64,
+{
+    batches: Vec<&'a (ShardId, Vec<u8>, T)>,
+    batches_idx: usize,
+    registers: &'a VecDeque<(ShardId, T)>,
+    register_idx: usize,
+}
+
+impl<'a, T> UnappliedIter<'a, T>
+where
+    T: Timestamp + Lattice + TotalOrder + StepForward + Codec64,
+{
+    fn new(
+        batches: &'a BTreeMap<usize, (ShardId, Vec<u8>, T)>,
+        registers: &'a VecDeque<(ShardId, T)>,
+    ) -> UnappliedIter<'a, T> {
+        UnappliedIter {
+            batches: batches.values().collect(),
+            batches_idx: 0,
+            registers,
+            register_idx: 0,
+        }
+    }
+}
+
+impl<'a, T: Timestamp + Lattice + TotalOrder + StepForward + Codec64> Iterator
+    for UnappliedIter<'a, T>
+{
+    type Item = (&'a ShardId, Unapplied<'a>, &'a T);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match (
+            self.batches.get(self.batches_idx),
+            self.registers.get(self.register_idx),
+        ) {
+            (Some((batch_id, batch, batch_ts)), Some((register_id, register_ts))) => {
+                if register_ts <= batch_ts {
```

Comment:
Always emitting registers before batches at an equal timestamp makes sense to me, let's leave a comment that this was on purpose

Reply:
I did it because it made intuitive sense, but I'm not actually sure that it matters. We know that for a single data shard we can't have both a register and a batch at the same timestamp. So it should be fine to apply all batches at a specific timestamp before all registers. All that just to say, I'm not actually sure what I'd say in the comment.

Reply:
Something like "Intentionally emit registers before batches at the same timestamp. This isn't necessary right now, but it would be if we ever add the ability to register and commit a data shard in a single op (and thus a single ts)."?

Reply:
I added something.